File: 第一章_免费试看/两串旋转练习题/Rotation.java
import java.util.*;
 
public class Rotation {
    public boolean chkRotation(String A, int lena, String B, int lenb) {
        if(A==null||B==null||lena<0||lenb<0){
            return false;
        }
        A=A+A;
        return kmpSearch(A,B)==-1?false:true;
    }
    public int [] getnext(String str){
        int len=str.length();
        int j=0;
        int k=-1;
        int []next=new int [len];
        next[0]=-1;
        while(j<len-1){
            if(k==-1||str.charAt(j)==str.charAt(k)){
                k++;
                j++;
                if(str.charAt(k)!=str.charAt(j)){
                    next[j]=k;
                }else{
                    next[j]=next[k];
                }
            }else{
                k=next[k];
            }
        }
        return next;
    }
    public int kmpSearch(String str1,String str2){
        int len1=str1.length();
        int len2=str2.length();
        int [] next=getnext(str2);
        int j=0,i=0;
        while(j<len1&&i<len2){
            if(i==-1||str1.charAt(j)==str2.charAt(i)){
                i++;
                j++;
            }else{
                i=next[i];
            }
        }
        return i==len2?j-i:-1;
    }
}

File: 第一章_免费试看/两串旋转练习题/Rotation.cpp
class Rotation {
public:
    bool chkRotation(string A, int lena, string B, int lenb) {
        if(lena!=lenb)
        {
            return false;
        }
        A=A+A;
        int res=kmpSearch(A,lena+lena,B,lenb);
        return res!=-1;
    }
 
    int kmpSearch(string A,int lena,string B,int lenb){
        int i=0,j=0;
 
        int *next=generateNext(B,lenb);
 
        while(i<lena&&j<lenb)
        {
            if(j==-1||A.at(i)==B.at(j))
            {
                i++;
                j++;
            }
            else
            {
                j=next[j];
            }
        }
        if(j==lenb)
        {
            return i-j;
        }
        else
        {
            return -1;
        }
    }
 
    int * generateNext(string B,int lenb){
        int j,k;
        j=0;k=-1;
        int * next=new int [lenb];
        next[0]=-1;
         
        while(j<lenb-1)
        {
            if(k==-1||B.at(k)==B.at(j))
            {
                k++;
                j++;
                if(B.at(k)!=B.at(j))
                {
                    next[j]=k;
                }
                else
                {
                    next[j]=next[k];
                }
            }
            else
            {
                k=next[k];
            }
        }
 
        return next;
    }
};